adr/ADR-33.md

```diff
@@ -0,0 +1,236 @@
+# Service API
+
+| Metadata | Value                 |
+| -------- | --------------------- |
+| Date     | 2022-11-23            |
+| Author   | @aricart              |
+| Status   | Partially Implemented |
+| Tags     | client                |
+
+## Context and Problem Statement
+
+Simplify the development of NATS micro-services.
+
+The design goal of the API is to reduce development to having a complexity
+similar to that of writing a NATS subscription, but then by using simple
+configuration, allow the specified metadata to allow for standardization of
+discovery and observability.
+
+## Design
+
+Service configuration relies on the following:
+
+- `name` - really the _kind_ of the service. Shared by all the services that
+  have the same name. This `name` cannot have any spaces or periods.
+- `version` - an opaque string identifying the version
+- `description` - a human-readable description about the service
+- `schema`:
+  - `request` - a string/url describing the format of the request payload
+  - `response` - a string/url describing the format of the response payload
+- `statusHandler` - an optional function that returns unknown data that can be
+  serialized as JSON. The handler will be provided the endpoint for which it is
+  building a `EndpointStats`
+- `endpoint` - a subject and a handler (effectively equivalent to a NATS
+  subscription)
+
+All services are created using a function called `addService()` where the above
+options are passed. The function returns an object/struct that represents the
+service. At a minimum the service is expected to offer functions/methods that
+allow:
+
+- a `stop(error?)` function that allows user code to stop the service.
+  Optionally this function should allow for an optional error. Stop should
+  always drain its service subscriptions.
+- `reset()` to reset any tracked metrics
+- `stats()` to return the status of the service
+- `done` a callback handler or promise where the framework can notify when its
+  subscriptions have stopped. Note that this is independent of the NATS
+  connection, and it should be possible to run multiple services under a single
+  connection.
+
+On startup a service is assigned an unique `id`. This `id` is used to
+distinguish different instances of the service and allow for a specific instance
+of the service to be addressed.
+
+### Discovery and Status
+
+Using the specified `name` and automatically generated `id` the service will
+automatically create a subscription to handle discovery and monitoring requests.
+
+The subject for discovery and requests is always composed of all capital
+letters. Prefixed by `$SRV`.
+
+The initial _verbs_ supported by the service include:
+
+- `PING`
```

Comment:
why do we have ping if it's just info?

Reply:
we are going to use `PING` for RTT - the payload now is just the `name` and `id` of the service.